compiler/rustc_parse/src/parser/item.rs

```diff
@@ -1849,23 +1852,15 @@ impl<'a> Parser<'a> {
         lo: Span,
         vis: Visibility,
         attrs: AttrVec,
-    ) -> PResult<'a, FieldDef> {
+    ) -> PResult<'a, Option<FieldDef>> {
```

Comment:
Can you make `Option<FieldDef>` into `Result<FieldDef, ErrorGuaranteed>`?

We should only ever return `Err` when a diagnostic is emitted.

Reply:
I think if we return `﻿Err` directly, the commit https://github.com/rust-lang/rust/commit/f25f9fe7d239dc866dbc0e522eb6011e6acb9806  is sufficient. Would it be advisable to revert it?

Reply:
No, we're not returning a `DiagnosticBuilder` here, we're just returning the `ErrorGuaranteed` token from emitting that error.

Reply:
The side effect of changing `﻿Option` to `﻿Result` involves the implementation of `﻿HasTokens` and `﻿HasAttrs` for `﻿Result`. While it may appear a bit unconventional, it seems remains acceptable.

Another aspect possibly requiring modification might be the return type of [parse_enum_variant](https://github.com/rust-lang/rust/blob/master/compiler/rustc_parse/src/parser/item.rs#L1461).

Reply:
I believe it was already like this. We should probably modify `PResult` to optionally have a `DiagnosticBuilder`, somehow.

Reply:
Please do not modify PResult 😅

Reply:
Also returning Err for the PResult variant would cause us to fail parsing -- I thought the whole point of this modification was to recover?

Reply:
> Also returning the Err variant for the PResult would cause us to fail parsing -- I thought the whole point of this modification was to recover?

That's what I was worried about, it would require much more changes. We could implement `Try` on a different `PResult`, but is that worth the effort? Probably not

I don't like having a nested `Result` though, we have two error conditions but one which we should essentially ignore; `Result<Result<T>>` doesn't convey this properly.

Reply:
Having two results represent two levels of state here -- the first Result is parsing failure, and the second is to carry an ErrorGuaranteed to make sure we don't ever accidentally drop fields. I think it's totally fine, if only a bit awkward.